### DISCUSSION TITLE
OpenAPI Enforcer v2 Features Discussion

### DISCUSSION THREAD

**Discussion by Gi60s** [byu-oit/openapi-enforcer#107] [General]
## 👋 Welcome!
 This is a place to discuss the features that you want to be included in version 2 of the OpenAPI enforcer. Some questions you might ask yourself are:

- How should version 2 be different from version 1?
- What could make version 2 better?
- What have you really liked about version 1 that you'd like to see stay in version 2?

Some guidelines for contributing:

- Please be respectful of others.
- No idea is dumb, but not all ideas will make it into the final product.
- We are all intelligent and we all have something to bring to the table. Don't be afraid to share.
- Be open-minded about other people's ideas.
- Vote on ideas.

To get started, comment below and tell us about what you think version 2 should or should not do. Limit one idea per post to enable easy voting. You can use this markdown example to get started:

```md
**Title of Suggestion**

The description of the suggestion goes here.
```

**Comment by Gi60s** [comment]
**TypeScript**

Version 2 is underway and is being written in TypeScript. I hope to have a branch up soon that will allow others to see where the code is at.

**Comment by Gi60s** [comment]
**Run on NodeJS and the Browser - Isomorphic / Universal**

It would also be nice if this library could run on NodeJS and the browser. In the browser it wouldn't be able to read the file system, but it could still fetch `$ref` values over the internet.

**Reply by Gi60s** [reply]
This feature is making it into v2 and is currently working.

**Reply by BrooksCunningham** [reply]
This could be super helpful when running the library as a WASM as well. I would love to test out the functionality if v2 is available for testing.

**Reply by Gi60s** [reply]
I haven't done anything in WASM myself yet. Are you saying that if it can run on the browser that it can be compiled to run in WASM?

Loading an OpenAPI document should work. Validating will work but likely with some false results (still adding tests). Additional functions have not yet been added to the components.

You can see the progress on the v2 branch. Sorry that we don't have any documentation yet. One breaking change of the new system is that loading and validating an OAS document can be separated. This will significantly improve performance for running a system in production.

**Reply by BrooksCunningham** [reply]
I am still learning about WASMs as well. However, from limited my understanding, if it works in the browser then it is easier to port into a WASM. There are certain libraries like "fs" that do not run on the browser since there is not an accessible file system. The link below has a good summary of the specific challenge with "fs".

https://dev.to/aspecto/make-your-npm-package-work-on-both-node-js-and-browser-30n7

The reason I am interested in this functionality is to use a WASM that may inspect HTTP requests and responses via an HTTP/HTTPS proxy. Then I would compare the HTTP request data and response data with a given site's openapi schema to validate the information conforms to the schema. Let me know if I may provide additional information. Thank you for the great open source work!

**Comment by Gi60s** [comment]
**OpenAPI Document Errors Include File Location and Line Number**

When loading from a document, whether off the file system or over the internet, include file location and line numbers for where errors exist in the OpenAPI v2 or v3 document.

**Reply by Gi60s** [reply]
This feature is making it into v2. It's already implemented and somewhat tested.

**Comment by jasonpai** [comment]
Any plans on releasing 2.0 with 3.1 support? Love the feature scope of your project, but need 3.1 validation support.

**Reply by Gi60s** [reply]
I'm glad that you asked. I do plan to release 3.1 support with version 2.0.

**Reply by jasonpai** [reply]
nice, is it still a long way? need help?

**Reply by Gi60s** [reply]
I've been thinking about how to respond to this for about a week now. It has made a lot of progress, but I think there is still quite a bit of work to be done too. As far as validating OpenAPI documents, that is mostly done but needs additional tests written.

I will probably take you up on that offer of helping out, although right now I'm refactoring some things so that will need to wait a bit.

Thanks for the offer of help.
